### ISSUE TITLE
Incorrect indentation for single line if/for/while/etc, multiline chaining statements etc

### ISSUE THREAD

**Issue by johanolofsson** [Microsoft/vscode#43244]
I am experiencing some weird indentation issues. Not using any plugins to prettify or similar.

if (condition) {
    return; // <-- correct indentation
}

if(condition)
return; // <-- incorrect indentation

Some ppl claim this should be avoided. I don't agree. Using pair of { } for simple early exits messes up code. Putting the return on the end of the same line can be hard to read if the condition is long.
This is not exclusive to return. Applies to all single lines. 

same problems apply to for, while, etc.

var transformedValues =
originalValues <-- incorrect should be indented one level.
.where(condition) <-- incorrect, should be indented two levels.
.select(transform); <-- incorrect, should be indented two levels. 

function someFunction() {
    callToSomeOtherFunction(
         variableWithLongNameWhichRequiresASeparateLine,
         anotherVariableWithLongNameWhichRequiresASeparateLine);
    }  <-- incorrect should not be indented.
    function ... <-- incorrect the rest of the file is indented.

**Comment by johanolofsson** [comment]
Thx for picking this up! Btw, it applies to typescript as well.

```
class Foo {
    public bar(x: number): void {
        if(x == 0)
        return; // <-- incorrect
    }
}
```

**Comment by Spongman** [comment]
also related, after typing, the code looks like this:

```javascript
if (condition)
return;
```

but after 'Format Document' the indentation is correct:
```javascript
if (condition)
  return;
```

but if you put the cursor after the `;` and hit `ENTER`, the cursor is incorrectly indented:

```javascript
if (condition)
  return;
  |  <-- cursor is here, should be in column zero.
```

**Comment by rickrankin** [comment]
This also applies C/C++/D, as well. Probably anything derived from C syntax.

**Comment by ahfeel** [comment]
Also applies to PHP, very very frustrating :( Anybody figured out a solution or an extension that fixes this ?

**Comment by johanolofsson** [comment]
None as far as I know, the fact that this has not yet been addressed is kind of weird. Seems to me like a very fundamental part of a decent code editor.

I'm starting to wonder if this is actually a case of opinion where the responsible developers simply don't agree with this way of avoiding { } and therefore just happens to forget about it? Microsoft, pls prove me wrong on this!

**Comment by stanislavdavid** [comment]
There is even bug with this code: 

```
if (condition) { some code in 1 line }
    next command wrongly intended
```

**Comment by hleVqq** [comment]
How is this taking so long to fix?

**Comment by Spongman** [comment]
@rebornix can you explain why you do not consider this a bug?

**Comment by Spongman** [comment]
here's another weirdness.

1) new file
2) set filetype to 'JS'
3) type `if(true)<ENTER>`

after hitting `<ENTER>` the cursor drops to line 2, column **1**. this is clearly wrong. the cursor at this point should be indented, and only un-indented if the next keystroke is a `{`.

![image](https://user-images.githubusercontent.com/1088194/48032083-27989100-e10b-11e8-9a03-e523d644c227.png)



however...

1) new file
2) hit `<ENTER>` several times
3) go to line 1
4) type `if(true)<ENTER>`

**now** the cursor ends up indented correctly:

![image](https://user-images.githubusercontent.com/1088194/48032055-09329580-e10b-11e8-948f-485329ec9f1d.png)

**Comment by fbricon** [comment]
So I opened #62198 for a related issue but it was closed as duplicate of this one (I'm not convinced this is the exact same issue though), so I'll repeat the specifics here:

Create a javascript file containing an if statement without brackets, one line executed if the if statement is valid, then 2 unrelated lines next

```js
var i = 1;
if (i === 1)
    console.log('log from if');

console.log('line one');
console.log('line two');
```
2. invert the last 2 lines with alt+up or alt+down
3. lines are indented under the if statement

![indentation-issue-when-moving-lines](https://user-images.githubusercontent.com/148698/47748736-e2cfae80-dc61-11e8-8a5b-b74708218fb7.gif)

And now for the really fun part: on Mac, the same issue exists in 1.28.2, BUT works correctly in  1.29.0-insider:
![oct-30-2018 16-24-12](https://user-images.githubusercontent.com/148698/47748929-51ad0780-dc62-11e8-8f15-0c8c87a5f174.gif)

We [found the issue](https://github.com/redhat-developer/vscode-java/pull/694#issuecomment-434366213) while trying to fix a [similar problem in vscode-java](https://github.com/redhat-developer/vscode-java/issues/692), after copying the [indentation rules](https://github.com/Microsoft/vscode/blob/41ddd41b71ea2bd5c6e49125beeb3e951f3d9c7c/extensions/typescript-language-features/src/features/languageConfiguration.ts#L17-L18) from the js/ts server.

**Comment by hleVqq** [comment]
@zhouhaoyu it still has weird behavior, like:
`function Test(x, y)
{
    if (x)
        return y;
    
    if (y)
    {
        
        }
}`
The setting also makes no difference in other languages, like C++.

**Comment by zhouhaoyu** [comment]
@hleVqq I guess there's a conflict between the auto indent feature and language-specific formatting then.
Hope this could be fixed soon...

**Comment by stanislavdavid** [comment]
Why this is labelled as feature request and bug label was removed. It is bug at the first glance

**Comment by fbricon** [comment]
The behavior I described in https://github.com/Microsoft/vscode/issues/43244#issuecomment-436351839 is still broken in VS Code 1.33.1 (and 1.34.0-exploration) but works fine in 1.34.0-insider O_o. So there's a proper behavior that's been there in the insider builds for months but never made it to the actual release

**Comment by d-cmyk** [comment]
Are assignees ever going to take ownership of this issue or will VS Code remain a broken editor forever?

**Comment by HyNek0** [comment]
I'm still looking for a fix too. This is really annoying and even basic editors do it right. If anyone finds an extension to fix this, please put it here too :/

**Comment by Prurite** [comment]
Any available fix now? I'm experiencing the same and this is really *annoying*.

**Comment by nfq-sb-tmp** [comment]
From the looks of it, this is never gonna be addressed. Though I'm unable to comprehend _why_.

**Comment by hleVqq** [comment]
From the looks of it, this is never gonna be fixed. Though I can't comprehend _why_.

**Comment by johanolofsson** [comment]
Yes, this is sad! Like I expressed in the beginning this is likely to be a case of opinion among a select few responsible for deciding what to do with this. And this is simply their way of enforcing that. But then again, shortly WASM will dominate the Web and we can all stop writing JS and go back to actually useful IDEs like #realVisualStudio.

**Comment by Prurite** [comment]
> Yes, this is sad! Like I expressed in the beginning this is likely to be a case of opinion among a select few responsible for deciding what to do with this. And this is simply their way of enforcing that. But then again, shortly WASM will dominate the Web and we can all stop writing JS and go back to actually useful IDEs like #realVisualStudio.

No this is not only the problem of JS... It affects all languages with this feature like C/C++ ( if I'm not misunderstanding you )

**Comment by johanolofsson** [comment]
You are correct that it is an issue for most, or perhaps all, languages in VSCode. My point being that there are imo better choices of IDEs for most other languages than JS/TS, and hopefully we will soon not have to use JS/TS as extensively as we do today due to the rise of WASM.

**Comment by svivian** [comment]
Just started using VS Code and I cannot believe this bug has existed for so long. This is language syntax 101... regardless of your opinion on optional braces they are optional in many languages so it's a valid way to write code. Does nobody on the dev team ever run into this issue, even occasionally?

**Comment by Spongman** [comment]
another weird javascript indentation bug:

````javascript


a(
	0,
	1, {
	a: "b",
}
)

````

**Comment by nunojsa** [comment]
> regardless of your opinion on optional braces they are optional in many languages so it's a valid way to write code

I completely agree with this. Some projects, like linux kernel even enforce this coding style on single line `if, for, etc...`. Hope to see a fix on this since it is rather annoying one...

**Comment by stanislavdavid** [comment]
lets wrote them this bug to survey in vscode which appeared to me today :-D, maybe it helps somewhere

**Comment by destructor465** [comment]
I use JS at work and C# at home and I'm using VS Code at work, recently I was thinking to moving from Visual Studio to VS Code (at home), I like how lightweight and simple it is. However this one little thing annoys me so much that everytime I experiecing it I lose all desire to do so.

**Comment by Spongman** [comment]
yet another trivial error:

````
return true ? {
} : {
	};
````

i don't understand why this is labeled `feature-request` and not `bug`.

**Comment by Prurite** [comment]
@rebornix @mjbvz take a look please

**Comment by ciabaros** [comment]
I've seen developers argue that this is not a bug, and others argue that you should use braces, and never single line indented statements.

The bizarre part is that every single other editor with auto indentation seems to do it right, including all versions of full Visual Studio, since beginning of time. I don't get it... why can't you just do what VS (full) does?

Yes, I can use another editor, and no, my problem isn't lack of effort by the devs for a free product, it's the mind blowing attitude that this is not a problem or not one worth fixing for like 5 years now. 

I'm used to VS Code and otherwise love it, but it has now become second nature to fix indentation after all single-line if/else/while/for/etc. -- why is this so hard to finally fix??

**Comment by Spongman** [comment]
It’s lazy program-management. It should be labeled as, and treated as a bug.

**Comment by FuriousBot1** [comment]
Could anyone please try to do some progress here? Since 2018 and it is such a simple detail, shouldn't be so difficult to solve, even less in an editor supported by Microsoft.

**Comment by yashbg** [comment]
Isn't there some extension or some other setting which could solve this issue?

**Comment by PrinceMatthew** [comment]
Seems like everybody has given up the hope of seeing this fixed. I'm moving to Jetbrains.

**Comment by Medallyon** [comment]
bump @rebornix @mjbvz 
Line after braceless if statement doesn't auto-indent

**Comment by Prurite** [comment]
2 yrs and a half have gone, even with an active PR they're still ignoring this :<

**Comment by g4borg** [comment]
is there any known workaround to have immediate indentation until this is finally merged? like with EditorConfig or similar?

(except patching manually with the PR, which in itself looks quite potent)

**Comment by eeegnu** [comment]
> is there any known workaround to have immediate indentation until this is finally merged? like with EditorConfig or similar?
> 
> (except patching manually with the PR, which in itself looks quite potent)

I was just working with this and I have a relatively simple solution here for the if cases https://stackoverflow.com/questions/69400313/how-to-make-vscode-indent-an-if-statement-without-brackets/69418298#69418298, though it gets broken when the official c/c++ debugger extension is enabled.

**Comment by hleVqq** [comment]
> > is there any known workaround to have immediate indentation until this is finally merged? like with EditorConfig or similar?
> > (except patching manually with the PR, which in itself looks quite potent)
> 
> I was just working with this and I have a relatively simple solution here for the if cases https://stackoverflow.com/questions/69400313/how-to-make-vscode-indent-an-if-statement-without-brackets/69418298#69418298, though it gets broken when the official c/c++ debugger extension is enabled.

It's missing other keywords (`else`, `while`, `for`, etc.), doesn't allow space after the keyword (must be `if(true)` and not `if (true)`) and doesn't unindent if you add a brace (`{`).

**Comment by eeegnu** [comment]
> > > is there any known workaround to have immediate indentation until this is finally merged? like with EditorConfig or similar?
> > > (except patching manually with the PR, which in itself looks quite potent)
> > 
> > 
> > I was just working with this and I have a relatively simple solution here for the if cases https://stackoverflow.com/questions/69400313/how-to-make-vscode-indent-an-if-statement-without-brackets/69418298#69418298, though it gets broken when the official c/c++ debugger extension is enabled.
> 
> It's missing other keywords (`else`, `while`, `for`, etc.), doesn't allow space after the keyword (must be `if(true)` and not `if (true)`) and doesn't unindent if you add a brace (`{`).

Yes it's definitely not a full solution, but since it took me ages to even get something responsive (I didn't think the official extension would break things), I figured it might prompt some discussion on this topic again. The other keywords are relatively simple to include, though else/else if may take their own logic; the second condition is fixable with another \s*. The brace I think would take an unindent rule, but those didn't work out well when I tried them.

**Comment by ciabaros** [comment]
> Indentation scopes in general seem to be completely neglected in vscode. It will indent for you when hitting enter within brackets, and it will maintain your current indentation, but that seems to be the full extent of its assistance. For example it will happily let you reduce indentation below the level of the containing block, rather than automatically wrap to the row above, which I can't imagine being desirable in any situation.

Yup, and fun fact: devs don't care

**Comment by ssigwart** [comment]
I submitted PR #136577 to resolve some of these issues.  I have comments on the pull request about which parts are fixed.

**Comment by ygoe** [comment]
@ssigwart Are you aware that there is already an open PR for this issue? And that looks pretty complete and ready. And just as ignored as yours. Why do people still try to help if help is obviously not appreciated here.

It's a shame that VS Code is now officially abandoned and receives no updates anymore (apart from those monthly version bumps that bring all sorts of new trouble). Are there any bugs fixed anymore? It's just ridiculous how Microsoft fails with this code editor that can't even edit code with the comfort of a much older plain-text editor like Notepad++ (or whatever you've used before).

Looks like it's finally time for a community fork of VS Code that simply merges all those ready bug fix PRs. That would make it a much better product. I think Microsoft is the wrong host for such a project, they simply have no interest in it.

**Comment by ssigwart** [comment]
@ygoe, I don't believe I knew there was an open PR.  I think I stopped reading the thread when it devolved into complaints about product management and therefore missed the PR.  However, I think it's valuable to have 2 PRs out there.  I took a look at #115454.   It looks like this will break the workflow of most users by adding a tab where it didn't use to.  I see that it will remove it if you type a brace, but that's still likely to cause complaints from people that always use braces. I have no backing for this, but I assume that Microsoft would be more likely to merge smaller changes than large ones.  So my PR fixes a small piece of this (deindenting after a single line `if`) that shouldn't break anyone's workflows.



> VS Code is now officially abandoned and receives no updates anymore

They do releases almost every month.

> Are there any bugs fixed anymore?

Yes.  I, as a community member, have 3 bug fixes and 2 feature request PRs that will go out in the August release.  So there is opportunity for community PRs to be merged.  Unfortunately, some PRs get no feedback whatsoever.  In those cases, I suggest that people upvote issues as indicated in https://github.com/microsoft/vscode/wiki/Issues-Triaging#up-voting-a-feature-request.

> Looks like it's finally time for a community fork of VS Code that simply merges all those ready bug fix PRs. That would make it a much better product.

There are pros and cons to that, but nothing is stopping that.  Personally, I'd rather not fragment the editor, which I think will hurt the editor long term, so I'm going to stick with VS Code and try to submit PRs for bugs that affect me or just random bugs Microsoft marks as `help-wanted` as a thank you for creating a free editor for me to use.  It's not perfect, but nothing ever is and for me, it's the best of the alternatives.

**Comment by ciabaros** [comment]
@ssigwart 
Yes, people get ranty when a black and white issue - that is solved just fine in virtually all other editors (including the full VS) - is pretty much called unreasonable or deeply problematic by the maintainers (within this issue's long history), for reasons that i cannot see as anything but anti single line brace-less structures opinionation at the core.

Yes, VS Code is free - and yes we appreciate the efforts to otherwise still keep it our favorite editor for many other positive reasons. But it's just surreal to see such push back about such a basic problem that has sadly trained thousands of developers manually tab after every single like if/else/for/etc. And to not see that and to that "we ignore because people rant" is just unprofessional -- why not really think about WHY people who otherwise love the product and respect your efforts turned to ranting out of frustration?

**Comment by ygoe** [comment]
@ciabaros Developers ignoring users' problems because they became "unfriendly" due to the issues the developers have caused, probably comes from the lack of social skills or empathy which is often observed among developers. Some project teams are worse than others. And teams tend to attract like-minded people, so the trouble is even self-reinforcing. Only few events are powerful and disruptive enough to change that.

BTW, I've tried out the auto-formatting option today since I read about it further up in this issue. Format on paste resolves the broken indenting of the following line (no other editor does that). Format on type resolves this very issue in .js files but not in script sections in .html files. And either of these settings annoys me when editing .scss files so much that I had to selectively turn this formatting on/off for certain file types.

So formatting is necessary to fix some situations in some file types, but formatting cannot be recommended in general because it can severely damage your code, too.

All in all, this product isn't mature. It just doesn't feel satisfying. Too many things haven't been properly thought through. And we're not talking about accessories, but the main and primary core feature: the code editor. If the devs were using this product themselves it would be a great chance to find and fix bugs before the public sees them. Or maybe they're just not aware or attentive (there are too many translations for that).

**Comment by quyykk** [comment]
Wow this bug is still open. There is an open PR. Please.

The format on type option doesn't seem to work for C++ files as well :confused:.

**Comment by DenNukem** [comment]
Come on, guys. Surely there is value in bringing VS Code in alignment with VS? That way people can easily move between the two - clearly a noble goal.

**Comment by me21** [comment]
👍

**Comment by aviden** [comment]
It's a shame that issues like this one remain open for years (and with two PRs available!). Unfortunately, I myself is not an expert in language features and VSCode internals to be able to adequately review them.

**Comment by Taschenbuch** [comment]
I can confirm this bug. This is super annoying when you write something like

```
if(isError)
  return;

  doSth() // and this lands on the same indention as the return though it should not. 
```

for some reasons "Format Documents" fixes it, but that is not a solution because while typing the code it will automatically mess up the idention like in the example above.

**Comment by parthasaradhie** [comment]
Is any maintainer  using vscode for development purpose?
If so they sure is encountered this bug a lot of times and yet this is still open ....
This is very annoying bug kindly look into this.

**Comment by aiday-mar** [comment]
Hi thanks for posting this issue, we have merged a PR to fix the indentation in the case of if/for/while statements in TypeScript/JavaScript that are followed by no brace. The line after should be indented. The same applies for arrow functions.

We are currently thinking about how to solve indentation for chained methods. This may require some more complex changes.

In the meantime, you may also turn on `format on save` or `format on type` from the settings. In most cases, this should give you correctly formatted code.

**Comment by me21** [comment]
Is it expected to work for C++ as well?

**Comment by aiday-mar** [comment]
Hi @me21 the fix that was merged is only for TypeScript/JavaScript. We can also make a corresponding fix for C++ and other languages (each language has its own language configuration file that needs to be edited). I will open an issue about this.

**Comment by ciabaros** [comment]
> Hi thanks for posting this issue, we have merged a PR to fix the indentation in the case of if/for/while statements in TypeScript/JavaScript that are followed by no brace. The line after should be indented. The same applies for arrow functions.
> 
> We are currently thinking about how to solve indentation for chained methods. This may require some more complex changes.
> 
> In the meantime, you may also turn on `format on save` or `format on type` from the settings. In most cases, this should give you correctly formatted code.

I think I speak for the community, when I say THANK YOU! It's a seemingly small, but extremely impactful issue. Can't wait till the VS Code update comes and we can try it out. All the best!

**Comment by aiday-mar** [comment]
Thank you @ciabaros, I appreciate it. I just realized that I did not propagate the change to the JavaScript language configuration file from the TypeScript language configuration file. Unfortunately this means the improvements I mentioned to JavaScript indentation will not be in this month's release like for TypeScript, but will appear in the release next month. If you wish to get the changes in as soon as possible you may download [VS Code Insiders](https://code.visualstudio.com/insiders/) which is updated on a daily basis with all the newest changes from the day before.

**Comment by ssigwart** [comment]
@aiday-mar, is there a setting to disable this? As I mentioned in the main description for PR #136577, always adding an extra tab is annoying because not all `if` statements are going to be single line `if`s.

**From PR:**
> while I like single line `if` statements without brackets, I don't want VS Code to assume all my `if` statements are going to be a single line for 2 reasons:
>
> If I do add brackets, it then looks like this:
> ```javascript
> if (1)
>   {
>     console.log('');
>   }
> ```
> I don't want to break other people's workflows. It seems more intuitive to hit tab once if you want to do a single line `if` > versus having to hit backspace if it were to assume it's going to be one line.

I added some screen records of the behavior for JavaScript in VS Code 1.88.0 vs Insider:

https://github.com/microsoft/vscode/assets/1426848/8f22e73d-3266-4288-8180-1702805cb157

https://github.com/microsoft/vscode/assets/1426848/8e974448-6564-408a-89d4-7eeae8d4ba5e

I much prefer the current behavior over Insiders. The same applies to Typescript, but this change has already made it into 1.88.0.

**Comment by me21** [comment]
Can't it be indented by default, but unindent if an opening brace is typed?

**Comment by ssigwart** [comment]
> Can't it be indented by default, but unindent if an opening brace is typed?

I was thinking of that as an option too. It's not my preference though. I'm hoping these can be added as settings so that if people want it auto-indented, they can enable the setting and if they want it to unindent when a brace is added, they can enable that setting.

**Comment by ciabaros** [comment]
> Can't it be indented by default, but unindent if an opening brace is typed?

I agree, we need this (ideally), or a way to turn this off altogether...

**Sadly**, _in its current state_ (without the above fix), this is much much worse than before, because: 
- For anyone (like me) who puts all their regular braces on new lines, this will now have a much bigger incorrect indentation impact, than the original lack of brace-less single line indentation.
- **If** this current new behavior is integrated into the "Format Document" and "Format Selection" options, then all my code gets broken from here on, and I think I would have to switch editors altogether.
- In any case, we'll have to get used to another even bigger quirk, in exchange.

While I'm thankful for the efforts to try and address this community request, in its current state, I think it needs to be pulled back, unless/until the un-indent brace fix above can be implemented.

**Comment by Taschenbuch** [comment]
This seems to be language specific. In Javascript/Typescript developers puts opening brace on the same line and not the next newline while in c# you put it on a newline. This becomes more difficult when people, like above, differ from this inofficial default code styles for the languages they use.

suggestion: 
Pressing enter after an if adds the indention. -> Then typing an opening brace '{' removes this indention again.
This way it will work for both c# and javascript and people who follow custom code styles.

**Comment by hleVqq** [comment]
This feature (though looks like more of an oversight to me) is needed for various C-like languages and you cannot assume that the opening brace is not going to go on a separate line.

This needs to work:

```
if (foo)
    bar();
```

This needs to work:

```
if (foo) 
{
    bar();
}
```

And this needs to (continue to) work:

```
if (foo) {
    bar();
}
```

**Comment by ssigwart** [comment]
> In order to fix this issue for those who want a different behavior, we can explore the feature of allowing users to define their own indentation rules within the settings to override those defined by the language configuration files. You could then define rules that better suit your needs. I will open an issue about this and discuss with colleagues.

Thanks, @aiday-mar. That sounds useful. One thing I was thinking that is kind of along these lines is that there could potentially be multiple language configuration files and VS Code would pick the appropriate one based on the settings the user picks. Though perhaps loading one config file, then manipulating it based on user setting would work too.

**Comment by Spongman** [comment]
@ssigwart no configuration is necessary to distinguish between the 3 examples that @hleVqq gave above.

This whole debacle can simply be solved like this:

when enter is pressed, if a new scope was created, indent accordingly. If the first character entered on the first, indented, otherwise-blank line is a ‘{‘, then un-indent that line first before inserting the ‘{‘.

**Comment by svivian** [comment]
> when enter is pressed, if a new scope was created, indent accordingly. If the first character entered on the first, indented, otherwise-blank line is a ‘{‘, then un-indent that line first before inserting the ‘{‘.

Yes, this is exactly it. Sublime Text and other editors already did it this way years (decades?) ago.

**Comment by aiday-mar** [comment]
Yes thank you for the clarification, I have opened a ticket here to track this feature : https://github.com/microsoft/vscode/issues/209822

**Comment by ssigwart** [comment]
> when enter is pressed, if a new scope was created, indent accordingly. If the first character entered on the first, indented, otherwise-blank line is a ‘{‘, then un-indent that line first before inserting the ‘{‘.

@Spongman, that is a good solution, but what I'm trying to point out is that it's not the ideal solution for everyone. For me, I could live with that over the behavior in https://github.com/microsoft/vscode/issues/43244#issuecomment-2041105031, but I personally don't like when the editor adds characters that don't belong. I have this urge to hit backspace first to get rid of the tab instead of assuming that the editor will remove it for me. I feel like I'm fighting the editor.

I'm not suggesting this gets implemented, but an equally valid solution to your unindenting suggestion would be to not indent and automatically indent if any character other than `{` is pressed.

My ideal solution is that no tab gets added and I can add the tab if needed. I know I'm probably in the minority here, so I'm not advocating for that to be the only option. I know people have different preferences, so the thing I'm advocating for is a setting to control the behavior.

**Comment by aiday-mar** [comment]
Hi @ssigwart thank you for your input, I understand your preferences. I have discussed the feature of allowing users to configure the indentation rules in the settings, and our team so far is leaning overall towards not providing it for several reasons:

1. It surfaces complexity that should ideally remain hidden
2. When the indentation rules are overridden this could cause unforeseeable indentation errors that the user did not take into account

If there are sufficient upvotes on the issue https://github.com/microsoft/vscode/issues/209805 we could reconsider allowing users to override the indentation rules in the settings.

In the meantime, you can also create a language extension that provides a language configuration file for your language of interest which contains the modified indentation rules. The advantage of developing an extension is that others can also benefit from the same indentation rules.

**Comment by ssigwart** [comment]
Thanks, @aiday-mar.

> In the meantime, you can also create a language extension that provides a language configuration file for your language of interest which contains the modified indentation rules. The advantage of developing an extension is that others can also benefit from the same indentation rules.

Yes, that was my fallback plan. I've done that for other languages before.

**Comment by FlowIT-JIT** [comment]
Adding indentation after flow control is a terrible ideer. Most people will immediately add curly brackets after a loop or if statement. Unfortunately indentation after flow control seems to be the new default behaviour in 1.88.x (https://github.com/microsoft/vscode/issues/116843) which results in an absolutely terrible editing experience.

**Comment by ygoe** [comment]
Did you mean "a terrible idea"?

Can you please explain what the issue is? From my understanding, an opening curly brace should unindent itself again automatically. At least that's the only way it makes sense if you want *any* automatic indentation. Every other way will cause pain somewhere.

**Comment by FlowIT-JIT** [comment]
> Did you mean "a terrible idea"?

Yes :-)

> Can you please explain what the issue is? From my understanding, an opening curly brace should unindent itself again automatically. At least that's the only way it makes sense if you want any automatic indentation. Every other way will cause pain somewhere.

If only adding curly brackets would outdent the code again, it would be fine. But it doesn't. Notice how I constantly have to fix my code in this video clip:

![VSCodeindentationbugs-ezgif com-video-to-gif-converter](https://github.com/microsoft/vscode/assets/10479064/79a556df-03ff-44ed-a39d-c48a87b75443)

**Comment by ciabaros** [comment]
@aiday-mar so, unfortunately this fix has introduced an even bigger problem, per comments above. How do we turn it off, until it is implemented to correctly unindent opening braces?

**Comment by ssigwart** [comment]
I noticed that the current implementation is a little inconsistent. If you hit Enter after `if` or `else if`, it will auto-indent. However, it does not indent after `else`.  Personally, I don't think any of them should indent, but if the first 2 do, I think `else` should as well to be consistent.

> In the meantime, you can also create a language extension that provides a language configuration file for your language of interest which contains the modified indentation rules. The advantage of developing an extension is that others can also benefit from the same indentation rules.

I created the `ssigwart.javascript-typescript-language-configuration` extension that will remove the auto-indenting after `if`, `for`, and `while` in JavaScript and Typescript.

**Comment by FlowIT-JIT** [comment]
> I created the `ssigwart.javascript-typescript-language-configuration` extension that will remove the auto-indenting after `if`, `for`, and `while` in JavaScript and Typescript.

So cool that you are trying to fix this mistake - much appreciated 💪😊
I do think they need to undo everything they changed in regards to indentation in 1.88 - it is completely broken, and it's beyond me that they havn't shipped a fix yesterday. This is a major bug breaking the editing experience.

**Comment by aiday-mar** [comment]
Hi @ssigwart thank you for pointing that out about the `else`. That is a bug and needs to be brought up to consistency with the other control flow statements. As I mentioned in the other threads, I believe the best experience to satisfy a majority of users would be to indent by default after an `if` statement, and when a curly brace is types, to outdent automatically. I believe this is how some other editors handle indentation too, and I will explore this further.

**Comment by FlowIT-JIT** [comment]
@aiday-mar Are you aware of the outdent issue documented in my video? Watch what happens when adding a line break after the `forEach(function(x)`
https://github.com/microsoft/vscode/issues/43244#issuecomment-2095440170
In this case curly brackets must indent the code.

**Comment by KyleKolander** [comment]
Yes, please fix this so it continues to work as it always did if you hit ENTER and then an opening curly brace. I spent a long time trying to figure out what I (or an installed extension) did to cause this to break.  I didn't even consider the possibility that a change like this would have been intentionally introduced.  It's one thing if it's just a bug, but this intentionally broke backward compatibility and deviates from the norm with other editors like VS.  Wow...

**Comment by aiday-mar** [comment]
Hi @FlowIT-JIT I have to look into that thank you.

After a braceless if statement the next line is indented, and when you type an opening curly brace, the line is outdented. This behavior is already in Insiders and will come out in the next stable release.

**Comment by 1337cookie** [comment]
Writing C in VSCode there's no auto-indentation after singe line while/if/for when not using brackets. 
I assume this just needs updates per language?

**Comment by RedCMD** [comment]
the current `onEnterRule`

![Image](https://github.com/user-attachments/assets/aa4353ac-26e8-416d-9071-4d796f99bd83)
